src/snowflake/snowpark/mock/_snowflake_data_type.py

```diff
@@ -134,6 +134,15 @@ def normalize_output_sf_type(t: DataType) -> DataType:
     return t
 
 
+def reset_nan_to_none(col_a, col_b, res_col):
+    # in pandas arithmetic operation, None are automatically convert to nan
+    if isinstance(res_col.sf_type.datatype, _NumericType):
+        for idx, (x, y) in enumerate(zip(col_a, col_b)):
+            if x is None or y is None:
+                res_col[idx] = None
+    return res_col
```

Comment:
I don' t think this works unless all the columns are object type. Otherwise pandas will just set the value back to nan.

Neither are object:
```
>>> a = pd.Series([1, 1, 1, None], index=['a', 'b', 'c', 'd'])
>>> b = pd.Series([1, None, 1, 1], index=['a', 'b', 'c', 'd'])
>>> res = a + b
>>> res
a    2.0
b    NaN
c    2.0
d    NaN
dtype: float64
>>> reset_nan_to_none(a, b, res)
a    2.0
b    NaN
c    2.0
d    NaN
dtype: float64
```
One side is object:
```
>>> a = pd.Series([1, 1, 1, None], index=['a', 'b', 'c', 'd'], dtype=object)
>>> b = pd.Series([1, None, 1, 1], index=['a', 'b', 'c', 'd'])
>>> res = a + b
>>> res
a    2.0
b    NaN
c    2.0
d    NaN
dtype: object
>>> reset_nan_to_none(a, b, res)
a     2.0
b     NaN
c     2.0
d    None
dtype: object
```

Both are object:
```
>>> a = pd.Series([1, 1, 1, None], index=['a', 'b', 'c', 'd'], dtype=object)
>>> b = pd.Series([1, None, 1, 1], index=['a', 'b', 'c', 'd'], dtype=object)
>>> res = a + b
>>> res
a      2
b    NaN
c      2
d    NaN
dtype: object
>>> reset_nan_to_none(a, b, res)
a       2
b    None
c       2
d    None
dtype: object
```